### ISSUE TITLE
A non-ChainerMN data parallel ImageNet example doesn't work well in a container including NVIDIA DALI.

### ISSUE THREAD

**Issue by lazykyama** [chainer/chainer#8063]
Hello, team!

I encountered a trivial issue that the ImageNet data parallel example (non-ChainerMN example; https://github.com/chainer/chainer/blob/v6.3.0/examples/imagenet/train_imagenet_data_parallel.py ) stops soon when the script runs in the environment where NVIDIA DALI is installed. According to the stack trace, the conflict of CUDA context initialization seems to cause this issue.

Conditions are below. I guess the number of affected users by this issue is very limited so it's not a critical problem. Please fix it if you team have enough time. (Or, I can make a PR. Please feel free to let me know if I should do so.)

* Conditions
  - Chainer version: chainer (6.3.0)
  - CuPy version: cupy-cuda100 (6.3.0)
  - OS/Platform: Ubuntu 18.04.2 LTS
  - CUDA/cuDNN version: 10.0.130 / 7.6.0
  - Python: 3.6.8
  - A container used in my test is built by this Dockerfile ( https://gist.github.com/lazykyama/92b235b8fe0f71ed7d000c9cbfee5732 ).
* Code to reproduce
  - `cd examples/imagenet/`
  - `python3 train_imagenet_data_parallel.py --arch resnet50 --devices @cupy:0 @cupy:1 --root / /path/to/imagenet/train_pair.txt /path/to/imagenet/val_pair.txt`
* Error messages, stack traces, or logs
  - Stack trace was shown like below.

```
Traceback (most recent call last):
  File "train_imagenet_data_parallel.py", line 149, in <module>
    main()
  File "train_imagenet_data_parallel.py", line 116, in main
    devices=devices)
  File "/usr/local/lib/python3.6/dist-packages/chainer/training/updaters/multiprocess_parallel_updater.py", line 138, in __init__
    'The CUDA context has been already initialized. '
RuntimeError: The CUDA context has been already initialized. MultiprocessParallelUpdater assumes the context is uninitialized. Please do not call CUDA API before MultiprocessParallelUpdater creates processes.
```

Workaround: making the timing to import DALI delayed like below.

```diff
diff --git a/examples/imagenet/train_imagenet.py b/examples/imagenet/train_imagenet.py
index bed9014fb..1557ea8c8 100755
--- a/examples/imagenet/train_imagenet.py
+++ b/examples/imagenet/train_imagenet.py
@@ -19,8 +19,6 @@ from chainer import training
 from chainer.training import extensions
 import chainerx

-import dali_util
-
 import alex
 import googlenet
 import googlenetbn
@@ -150,6 +148,8 @@ def main():
     # Load the mean file
     mean = np.load(args.mean)
     if args.dali:
+        import dali_util
+
         if not dali_util._dali_available:
             raise RuntimeError('DALI seems not available on your system.')
         if not isinstance(device, chainer.backend.cuda.GpuDevice):
```

**Comment by hvy** [comment]
Thanks for the detailed report, we really appreciate it. The current separation is not that clean I think, importing the whole `train_imagenet.py` in `train_imagenet_data_parallel.py` just to reuse `PreprocessDataset`. This is what's causing the problem. A cleaner approach would be to extract this class in a separate submodule and import this submodule in both training scripts, decoupling them.

What do you think? You are more than welcome to do this, or I could otherwise.

**Comment by lazykyama** [comment]
Thanks for your checking. I understand the situation and background well. I totally agree your approach is better. Separating a dataset class from main code to another (e.g., `dataset_util.py`) is a best solution.

Of course, I can try to make a PR to fix this issue. Please give me a few days.

**Comment by lazykyama** [comment]
@hvy and team,
Sorry for my too late action... Just making a PR. Please check it. (To be honest, this is my first PR to public repo. Please let me know if there are some mistakes...!)

BTW, during testing my change, I found another issue on running `train_imagenet_data_parallel.py` and Chainer v7.0.0b3. Will make a new issue later.

**Comment by hvy** [comment]
This issue is already resolved by #8127. Closing.
